modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java

```diff
@@ -583,28 +583,34 @@ static PercolateQuery.QueryStore createStore(MappedFieldType queryBuilderFieldTy
             return docId -> {
                 if (binaryDocValues.advanceExact(docId)) {
                     BytesRef qbSource = binaryDocValues.binaryValue();
-                    try (InputStream in = new ByteArrayInputStream(qbSource.bytes, qbSource.offset, qbSource.length)) {
-                        try (
-                            StreamInput input = new NamedWriteableAwareStreamInput(
-                                new InputStreamStreamInput(in, qbSource.length),
-                                registry
-                            )
-                        ) {
-                            input.setTransportVersion(indexVersion.transportVersion);
-                            // Query builder's content is stored via BinaryFieldMapper, which has a custom encoding
-                            // to encode multiple binary values into a single binary doc values field.
-                            // This is the reason we need to first need to read the number of values and
-                            // then the length of the field value in bytes.
-                            int numValues = input.readVInt();
-                            assert numValues == 1;
-                            int valueLength = input.readVInt();
-                            assert valueLength > 0;
-                            QueryBuilder queryBuilder = input.readNamedWriteable(QueryBuilder.class);
-                            assert in.read() == -1;
-                            queryBuilder = Rewriteable.rewrite(queryBuilder, context);
-                            return queryBuilder.toQuery(context);
+                    try (
+                        InputStream in = new ByteArrayInputStream(qbSource.bytes, qbSource.offset, qbSource.length);
+                        StreamInput input = new NamedWriteableAwareStreamInput(new InputStreamStreamInput(in, qbSource.length), registry)
+                    ) {
+                        // Query builder's content is stored via BinaryFieldMapper, which has a custom encoding
+                        // to encode multiple binary values into a single binary doc values field.
+                        // This is the reason we need to first need to read the number of values and
+                        // then the length of the field value in bytes.
+                        int numValues = input.readVInt();
+                        assert numValues == 1;
+                        int valueLength = input.readVInt();
+                        assert valueLength > 0;
+
+                        TransportVersion transportVersion;
+                        if (indexVersion.before(Version.V_8_8_0)) {
+                            transportVersion = TransportVersion.fromId(indexVersion.id);
+                        } else {
+                            transportVersion = TransportVersion.fromId(input.readInt());
```

Comment:
I think this is going to be confusing given there is `TransportVerison.readVersion`, which someone might come along here and change to without noticing the difference of Int vs VInt. Other than the cases where it's a chicken-and-egg problem (like handshake?), could we change `TransportVersion.readVersion` to have conditional logic to go from vint -> int? And in the edge case of needing to still read vint for handshake, change that case to read vint directly and call fromId.

What I suggest here is to change back to use `TransportVersion.readVersion` in this PR, and then separately we can think through conditionally changing the readVersion implementation.

Reply:
What would the read be conditional on? The transport version of the stream?